### ISSUE TITLE
RyuJIT: Bad codegen with multiple field assignments

### ISSUE THREAD

**Issue by jakobbotsch** [dotnet/coreclr#21231]
The example is:
```csharp
// Debug: Outputs 2
// Release: Outputs 1
struct S0
{
    public sbyte F0;
    public S0(sbyte f0): this()
    {
        F0 = f0;
    }
}

struct S1
{
    public S0 F0;
    public ushort F1;
    public S1(S0 f0): this()
    {
        F0 = f0;
    }
}

public class Program
{
    public static void Main()
    {
        var vr22 = new S1[]{new S1(new S0(1))};
        S1 vr26;
        vr26.F0 = vr22[0].F0;
        vr22[0].F0.F0 += vr22[0].F0.F0;
        vr26.F0 = vr22[0].F0;
        System.Console.WriteLine(vr26.F0.F0);
    }
}
```

Disassembly is:
```asm
       E841014F5F           call     CORINFO_HELP_NEWARR_1_VC
       B901000000           mov      ecx, 1
       33D2                 xor      edx, edx
       89542428             mov      dword ptr [rsp+28H], edx
       89542428             mov      dword ptr [rsp+28H], edx
       488D542428           lea      rdx, bword ptr [rsp+28H]
       880A                 mov      byte  ptr [rdx], cl
       8B4C2428             mov      ecx, dword ptr [rsp+28H]
       894810               mov      dword ptr [rax+16], ecx
       480FBE4810           movsx    rcx, byte  ptr [rax+16]
       884C2430             mov      byte  ptr [rsp+30H], cl
       4883C010             add      rax, 16
       480FBE10             movsx    rdx, byte  ptr [rax]
       03D2                 add      edx, edx
       8810                 mov      byte  ptr [rax], dl
       884C2430             mov      byte  ptr [rsp+30H], cl
       480FBE4C2430         movsx    rcx, byte  ptr [rsp+30H]
       E828ECFFFF           call     System.Console:WriteLine(int)
```
It seems the second assignment to `vr26.F0` has been missed, and RyuJIT then gets a stale value.

cc @briansull

**Comment by briansull** [comment]
This is a similar issue in ValueNumbering to  #18259
I was looking into it yesterday

**Comment by briansull** [comment]
Fixed with PR #21231

**Comment by AndyAyersMS** [comment]
So this test case didn't hit the assert, but still ended up with bad code?

I think we should add the code above as a regression test.

**Comment by briansull** [comment]
OK will add

**Comment by briansull** [comment]
I am reopening this issue.
This test failure depends upon the C# compiler that you compile the test case with.
It still fails using a Roslyn based C# compiler.

**Comment by briansull** [comment]
This test case is passing on CoreCLR 3.0
Recent changes in struct field promotion for CoreCLR has fixed this for the test case.

I will check in the test case.

**Comment by briansull** [comment]
Resolved with PR #23109
